### PR TITLE
Include llm training rules

### PR DESCRIPTION
**Pull request by ShriyaRishab**
High level list of rule changes
1. Minimum number of GPUs needed to train an llm reference
2. LLM metadata preprocessing need not be timed
3. Add quality metric, dataset, model details, hyperparameters and constraints
4. Add note on fixed learning rate and restrictions on hparams search
5. RCP generation exception and timeline

### CODE REVIEW

File: training_rules.adoc [hunk 1]
```diff
@@ -80,6 +82,7 @@ The benchmark suite consists of the benchmarks shown in the following table.
 | |Object detection (heavy weight) |COCO
 |Language |Speech recognition |LibriSpeech
 | |NLP |Wikipedia 2020/01/01
+| |Large language model |C4 3.0.1
```
Comment (sgpyc) [inline training_rules.adoc]: "C4 3.0.1" -> "C4/en/3.0.1"?

**Comment by nv-rborkar** [comment]
Training WG meeting 02/09: @itayhubara @sgpyc assigned for review

File: training_rules.adoc [hunk 2]
```diff
@@ -277,6 +281,19 @@ The MLPerf verifier scripts checks all hyperparameters except those with names m
  |dlrm |sgd |sgd_opt_base_learning_rate |unconstrained |learning rate at the start of poly decay |See PR (From Intel and NV, TODO Link)
  |dlrm |sgd |sgd_opt_learning_rate_decay_poly_power |2 |power of the poly decay |See PR (From Intel and NV, TODO Link)
  |dlrm |sgd |sgd_opt_learning_rate_decay_steps |unconstrained |the step at which you reach the end learning rate |See PR (From Intel and NV, TODO Link)
+ |llm |adam |global_batch_size |unconstrained |batch size in sequences |See PR (From NV and Google, TODO Link)
+ |llm |adam |opt_adam_beta_1 |0.9 |adam beta1 |See PR (From NV and Google, TODO Link)
+ |llm |adam |opt_adam_beta_2 |0.95 |adam beta2 |See PR (From NV and Google, TODO Link)
+ |llm |adam |opt_adam_epsilon |0.1 |adam epsilon |See PR (From NV and Google, TODO Link)
```
Comment (sgpyc) [inline training_rules.adoc]: opt_adam_epsilon = 1e-8

File: training_rules.adoc [hunk 3]
```diff
@@ -277,6 +281,19 @@ The MLPerf verifier scripts checks all hyperparameters except those with names m
...
+ |llm |adam |opt_gradient_clip_norm |1.0 |Gradients are clipped above this norm threshold. |See PR (From NV and Google, TODO Link)
+ |llm |adam |sequence_length |2048 |sequence length |See PR (From NV and Google, TODO Link)
+ |llm |adam |opt_weight_decay |0.1 |weight decay |See PR (From NV and Google, TODO Link)
+ |llm |adam |opt_gradient_accumulation_steps |unconstrained |Numer of fwd/bwd steps between optimizer step. |See PR (From NV and Google, TODO Link)
```
Comment (sgpyc) [inline training_rules.adoc]: Every benchmark has this value, and the name is "gradient_accumulation_steps".
Maybe we can skip it here?

File: training_rules.adoc [hunk 4]
```diff
@@ -277,6 +281,19 @@ The MLPerf verifier scripts checks all hyperparameters except those with names m
...
+ |llm |adam |opt_learning_rate_warmup_steps |ceil(265 * 1536 / global_batch_size) |steps taken for linear warmup during initial checkpoint generation. This only affects the learning rate curve in the benchmarking region. |See PR (From NV and Google, TODO Link)
+ |llm |adam |opt_learning_rate_decay_steps |ceil(108600 * 1536 / global_batch_size) |learning rate cosine decay from (opt_learning_rate_warmup_steps + 1,opt_learning_rate_decay_steps]. |See PR (From NV and Google, TODO Link)
```
Comment (sgpyc) [inline training_rules.adoc]: May be add "the step at which the end of the cosine learning rate curve is reached."?

File: training_rules.adoc [hunk 5]
```diff
@@ -392,6 +431,7 @@ CLOSED: The same quality measure as the reference implementation must be used. T
 |       |Object detection (heavy weight) |Mask R-CNN|Every 1 epoch 
 |Language|Speech recognition |RNN-T|Every 1 epoch
 |        |NLP |BERT| eval_interval_samples=FLOOR(0.05*(230.23*GBS+3000000), 25000), skipping 0
+|        |large Language Model |GPT-3| Every 24576 samples. CEIL(24576 / global_batch_size) if 24576 is not divisible by GBS
```
Comment (sgpyc) [inline training_rules.adoc]: "samples" => "sequences"?

**Comment by itayhubara** [comment]
Looks good to me I would consider adding that dropout is zero: attention-dropout 0.0, hidden-dropout 0.0

File: training_rules.adoc (hunk 3 above)
Comment (ShriyaRishab) [reply training_rules.adoc]: I think that it is still good to explicitly report the value even though it can be computed manually.

File: training_rules.adoc [hunk 6]
```diff
@@ -362,6 +379,27 @@ The MLPerf verifier scripts checks all hyperparameters except those with names m
...
+ |1536 |2.0e-5
+ |2048 |2.0e-5
+ |3072 |2.0e-5
+ |4096 |3.0e-5
+ |8192 |3.0e-5
+|===
+
+* GBS<1536 or GBS>8192 - new RCP needs to be generated, reach out to the task force
+* GBS [1536,3072] - opt_base_learning_rate=2.0e-5
+* For (3073,4096) - opt_base_learning_rate=2.0e-5 or opt_base_learning_rate=3.0e-5
```
Comment (sgpyc) [inline training_rules.adoc]: `(3073, 4096)` => `(3072, 4096)`

File: training_rules.adoc (hunk 5 above)
Comment (sgpyc) [inline training_rules.adoc]: `samples` => `sequences`

File: training_rules.adoc [hunk 7]
```diff
@@ -499,6 +542,7 @@ To extract submission convergence points, logs should report epochs as follows.
 
 | RN50 | Epoch
 | BERT | Training sample (integer)
+| LLM | Training sample starting from 0 (integer)
```
Comment (sgpyc) [inline training_rules.adoc]: `sample` => `token` ?

File: training_rules.adoc (hunk 5 above)
Comment (ShriyaRishab) [reply training_rules.adoc]: Fixed in commit: e996e70

File: training_rules.adoc [hunk 8]
```diff
@@ -499,6 +543,7 @@ To extract submission convergence points, logs should report epochs as follows.
 
 | RN50 | Epoch
 | BERT | Training sample (integer)
+| GPT3 | Training sample starting from 0 (integer)
```
Comment (sgpyc) [inline training_rules.adoc]: `sample` => `token` ?

**Review by sgpyc — CHANGES_REQUESTED** [review]
Looks good after comments are resolved.

**Comment by johntran-nv** [comment]
Training WG approved today.
